base/array.jl

```diff
@@ -668,10 +682,7 @@ function collect_to!(dest::AbstractArray{T}, itr, offs, st) where T
             @inbounds dest[i] = el::T
             i += 1
         else
-            R = promote_typejoin(T, typeof(el))
-            new = similar(dest, R)
-            copyto!(new,1, dest,1, i-1)
-            @inbounds new[i] = el
+            new = setindex_widen!(dest, el, i)
```

Comment:
Is there a way to avoid the duplication of the above `if el isa T || ...` part? (And same in `grow_to!`.)

Reply:
They aren't strictly necessary but I selfishly wanted it duplicated. It will be useful to have later for ModelArrays (where first you test whether the eltype of the whole row matches (which it should it type stable code) and then test each row to see which ones don't match (as an optimization). Essentially, I'll be running `setindex_widen!.(dests, els, i)`

Reply:
Now I'm feeling a bit guilty, though. I could also just implement my own `setindex_maybe_widen!` down the line

Reply:
Ok I removed the duplicated code